### DISCUSSION TITLE
Enhancing FinOps Toolkit: Calculating Realized Cost Savings at Billing Profile Level

### DISCUSSION THREAD

**Discussion by GregorWohlfarter** [microsoft/finops-toolkit#1064] [Ideas]
**Customer Challenge:**

A customer is struggling to accurately calculate cost savings from their cost optimization initiatives at the billing account level. They have a small team dedicated to various optimization tasks, such as right-sizing, storage optimization, and purchasing reservations and savings plans. The challenge is to provide detailed savings data per month, quarter, and year for each of their 20 billing profiles. The current tools, like Azure Advisor Cost Optimization Score, do not offer the granularity needed, and accumulating these numbers is complex due to the evolving infrastructure.

**Proposed Solutions and Suggestions:**

1. **Tracking Implemented Recommendations:**
   - **Approach:** Identify which specific recommendations were implemented by either noting when recommendations disappear from the system (risky to assume) or by logging remediation actions.
   - **Drawback:** This requires historical data, which is not directly available via Azure Advisor recommendations, making it challenging to accurately track and verify the implementation of recommendations over time.

2. **Calculating Savings:**
   - **Approach:** Compute the savings by comparing the costs of the remediated resources from one month ago to the present. For reservation recommendations, compare the overall effective cost of the reserved SKU, considering the same usage quantities, from one month ago to today.
   - **Drawback:** Accumulating realized savings from both reservations and savings plans can be complex. Additionally, calculating these savings at the billing profile level must account for the elasticity of a dynamically evolving infrastructure, such as autoscaling databases and compute clusters.

3. **Scoping Savings to Billing Profiles:**
   - **Approach:** Determine the billing profile scope of the remediation by examining the billing data, as each billing line in the cost details contains the billing profile ID.
   - **Drawback:** While this method can help scope realized savings to specific billing profiles, it may still be challenging to accurately calculate savings due to the dynamic nature of the infrastructure.

**Discussion Points:**

- How can we best integrate these solutions into the existing FinOps toolkit?
- What new features might need to be developed to support this functionality?
- Are there any additional suggestions or feedback from the community on how to approach this challenge?

**Comment by helderpinto** [comment]
Thanks for raising this interesting discussion, @GregorWohlfarter. We are planning to build an historical data set of Advisor (and custom) recommendations as part of FinOps hubs. This feature is currently being developed and should be available in one of the upcoming monthly releases. Correlating historical recommendations data with the FOCUS data set would help addressing the problem you describe, but I understand there is a lot to consider besides simply correlating data sets.

**Reply by flanakin** [reply]
I was thinking the same. We should create an issue to track this.

@GregorWohlfarter we'd love to have your help in building this out 🙂

**Comment by lundejd** [comment]
We currently have measures that tell us month to month how we are doing with things like RIs, savings plans, and our EA discount vs. public cost and are just using the data provided. 

One of my FinOps guys developed his own PowerShell scripts to analyze servers, sql, storage etc. at a more granular level which has really helped us save and be confident about it vs. what Microsoft will tell you. He uses this graph as the monthly test of how things are progressing. 

It would be great to see more of this in the FinOps hub so happy to hear that.

This is our server one, we have one for storage, sql, app services and a few others on the way. (amounts blurred for privacy of course but you get the idea)

<img width="1588" alt="Screenshot 2024-10-01 at 4 15 10 PM" src="https://github.com/user-attachments/assets/f2833d62-a7bf-4494-b433-16f09a9c8d78">

One huge problem with this is the Power BI file that generates this pulls data for almost 12 hours. I won't run in Power BI online and from what I can see would't probably make it past one month in Fabric due to their memory limitations.

**Reply by flanakin** [reply]
I'm very curious what the script is looking at. We could add something like that to the toolkit, if it'd be more broadly applicable, but that would likely be a challenge if it takes a long time to run. We could also use it as a baseline for what data to pull into hubs.

Once we have Azure Data Explorer, this should be a fairly simple thing to build. The challenge we have today is using raw data in storage.
